File: langs/go/masking.go
package dotlyte

import (
	"fmt"
	"regexp"
	"strings"
)

// REDACTED is the replacement string for sensitive values.
const REDACTED = "[REDACTED]"

// sensitivePatterns matches common secret key names.
var sensitivePatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?i)password`),
	regexp.MustCompile(`(?i)secret`),
	regexp.MustCompile(`(?i)token`),
	regexp.MustCompile(`(?i)api[_\-]?key`),
	regexp.MustCompile(`(?i)private[_\-]?key`),
	regexp.MustCompile(`(?i)access[_\-]?key`),
	regexp.MustCompile(`(?i)auth`),
	regexp.MustCompile(`(?i)credential`),
	regexp.MustCompile(`(?i)connection[_\-]?string`),
	regexp.MustCompile(`(?i)dsn`),
	regexp.MustCompile(`(?i)encryption[_\-]?key`),
	regexp.MustCompile(`(?i)signing[_\-]?key`),
	regexp.MustCompile(`(?i)certificate`),
}

// BuildSensitiveSet returns a set of keys that should be redacted,
// combining auto-detected keys with schema-declared sensitive keys.
func BuildSensitiveSet(data map[string]any, schemaKeys []string) map[string]bool {
	set := make(map[string]bool)

	// Add schema-declared keys
	for _, k := range schemaKeys {
		set[k] = true
	}

	// Auto-detect from key names
	keys := flattenKeys(data, "")
	for _, key := range keys {
		for _, pat := range sensitivePatterns {
			if pat.MatchString(key) {
				set[key] = true
				break
			}
		}
	}

	return set
}

// RedactMap returns a deep copy of the data map with sensitive values replaced by REDACTED.
func RedactMap(data map[string]any, sensitiveKeys map[string]bool) map[string]any {
	return redactMapInner(data, sensitiveKeys, "")
}

func redactMapInner(data map[string]any, sensitiveKeys map[string]bool, prefix string) map[string]any {
	result := make(map[string]any, len(data))
	for k, v := range data {
		fullKey := k
		if prefix != "" {
			fullKey = prefix + "." + k
		}

		if sensitiveKeys[fullKey] {
			result[k] = REDACTED
			continue
		}

		if m, ok := v.(map[string]any); ok {
			result[k] = redactMapInner(m, sensitiveKeys, fullKey)
		} else {
			result[k] = v
		}
	}
	return result
}

// FormatRedacted formats a value as a partially visible redacted string.
// Shows first 2 chars and masks the rest: "ab****"
func FormatRedacted(value string) string {
	if len(value) <= 4 {
		return strings.Repeat("*", len(value))
	}
	return value[:2] + strings.Repeat("*", len(value)-2)
}

// CompilePatterns converts glob-like patterns (e.g., "*_KEY", "DATABASE_*")
// to compiled regular expressions. Supports * as a wildcard for any characters.
// Returns an error if any pattern is invalid.
func CompilePatterns(patterns []string) ([]*regexp.Regexp, error) {
	compiled := make([]*regexp.Regexp, 0, len(patterns))
	for _, p := range patterns {
		// Escape regex special chars, then convert \* back to .*
		escaped := regexp.QuoteMeta(p)
		escaped = strings.ReplaceAll(escaped, `\*`, ".*")
		re, err := regexp.Compile("(?i)^" + escaped + "$")
		if err != nil {
			return nil, fmt.Errorf("dotlyte: invalid pattern '%s': %w", p, err)
		}
		compiled = append(compiled, re)
	}
	return compiled, nil
}

// BuildSensitiveSetWithPatterns builds a sensitive key set using custom glob
// patterns in addition to schema-declared sensitive keys and auto-detection.
// Patterns like "*_KEY", "*_SECRET", "DATABASE_*" match against key names.
func BuildSensitiveSetWithPatterns(keys []string, patterns []string, schemaSensitive map[string]bool) (map[string]bool, error) {
	result := make(map[string]bool)

	// Add schema-declared keys
	for k := range schemaSensitive {
		result[k] = true
	}

	// Compile custom patterns
	compiled, err := CompilePatterns(patterns)
	if err != nil {
		return nil, err
	}

	for _, key := range keys {
		// Check against custom patterns
		leaf := key
		if idx := strings.LastIndex(key, "."); idx >= 0 {
			leaf = key[idx+1:]
		}
		for _, re := range compiled {
			if re.MatchString(leaf) || re.MatchString(key) {
				result[key] = true
				break
			}
		}
		// Also check built-in patterns
		for _, pat := range sensitivePatterns {
			if pat.MatchString(leaf) {
				result[key] = true
				break
			}
		}
	}

	return result, nil
}

// AuditAccessFunc is a callback invoked when a sensitive key is accessed.
// The key parameter is the config key name, and context indicates the
// access context (always "server" in Go).
type AuditAccessFunc func(key, context string)

// AuditWrapper wraps a data map and fires an audit callback whenever a
// key in the sensitive set is accessed via Get.
type AuditWrapper struct {
	data          map[string]interface{}
	sensitiveKeys map[string]bool
	onAccess      AuditAccessFunc
}

// CreateAuditWrapper creates an AuditWrapper that triggers onAccess whenever
// a sensitive key is read. This is useful for SOC2/HIPAA compliance audit logging.
func CreateAuditWrapper(data map[string]interface{}, sensitiveKeys map[string]bool, onAccess AuditAccessFunc) *AuditWrapper {
	return &AuditWrapper{
		data:          data,
		sensitiveKeys: sensitiveKeys,
		onAccess:      onAccess,
	}
}

// Get retrieves a value by key. If the key is in the sensitive set,
// the audit callback is triggered before the value is returned.
func (aw *AuditWrapper) Get(key string) interface{} {
	if aw.sensitiveKeys[key] && aw.onAccess != nil {
		aw.onAccess(key, "server")
	}
	return aw.data[key]
}

File: langs/javascript/src/masking.ts
/**
 * Sensitive value masking and redaction for DOTLYTE.
 *
 * Auto-detects keys that contain sensitive data and redacts them in
 * toString(), toJSON(), and toObjectRedacted(). Actual values remain
 * accessible via get() and require().
 *
 * Auto-sensitive patterns (case-insensitive):
 *   secret, password, passwd, token, key, credential, auth, api_key,
 *   apikey, private, certificate, cert
 */

const SENSITIVE_PATTERNS = [
  /secret/i,
  /password/i,
  /passwd/i,
  /token/i,
  /\bkey\b/i,
  /credential/i,
  /\bauth\b/i,
  /api[_-]?key/i,
  /private/i,
  /certificate/i,
  /\bcert\b/i,
  /\bdsn\b/i,
  /\bsalt\b/i,
  /\bhash\b/i,
  /\bsigning/i,
  /\bencrypt/i,
  /connection[_-]?string/i,
];

/** The mask string used to replace sensitive values. */
export const REDACTED = "***REDACTED***";

/**
 * Check if a key name is auto-detected as sensitive.
 */
export function isAutoSensitive(key: string): boolean {
  return SENSITIVE_PATTERNS.some((p) => p.test(key));
}

/**
 * Build the complete set of sensitive keys from:
 *  1. Schema-defined sensitive: true
 *  2. Auto-detected patterns
 *  3. minus explicit sensitive: false overrides
 *
 * @param allKeys      — all config keys (flat, dot-notation)
 * @param schemaSensitive — keys explicitly marked sensitive in schema
 * @param schemaNotSensitive — keys explicitly marked sensitive: false
 */
export function buildSensitiveSet(
  allKeys: string[],
  schemaSensitive: Set<string> = new Set(),
  schemaNotSensitive: Set<string> = new Set(),
): Set<string> {
  const result = new Set(schemaSensitive);

  for (const key of allKeys) {
    if (schemaNotSensitive.has(key)) continue;
    // Check the leaf part of the key (after last dot)
    const leaf = key.includes(".") ? key.split(".").pop()! : key;
    if (isAutoSensitive(leaf)) {
      result.add(key);
    }
  }

  return result;
}

/**
 * Deep-redact sensitive values in a config object.
 *
 * @param data     — the config data
 * @param sensitive — set of sensitive keys (dot-notation)
 * @param prefix   — current key prefix for recursion
 * @returns New object with sensitive values replaced by REDACTED.
 */
export function redactObject(
  data: Record<string, unknown>,
  sensitive: Set<string>,
  prefix = "",
): Record<string, unknown> {
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(data)) {
    const fullKey = prefix ? `${prefix}.${key}` : key;

    if (sensitive.has(fullKey)) {
      result[key] = REDACTED;
    } else if (value !== null && typeof value === "object" && !Array.isArray(value)) {
      result[key] = redactObject(value as Record<string, unknown>, sensitive, fullKey);
    } else {
      result[key] = value;
    }
  }

  return result;
}

/**
 * Format a config object as a human-readable string with redaction.
 */
export function formatRedacted(
  data: Record<string, unknown>,
  sensitive: Set<string>,
): string {
  const redacted = redactObject(data, sensitive);
  return JSON.stringify(redacted, null, 2);
}

/* ──────── Configurable Sensitive Patterns ──────── */

/**
 * Convert glob-like patterns (e.g., "*_KEY", "DATABASE_*") to regex patterns.
 *
 * Supports:
 *   - `*` → match anything
 *   - Literal string matching (case-insensitive)
 *
 * @param patterns — array of glob patterns
 * @returns Array of compiled RegExp patterns
 */
export function compilePatterns(patterns: string[]): RegExp[] {
  return patterns.map((p) => {
    const escaped = p.replace(/[.*+?^${}()|[\]\\]/g, "\\$&").replace(/\\\*/g, ".*");
    return new RegExp(`^${escaped}$`, "i");
  });
}

/**
 * Build a sensitive key set using custom patterns.
 *
 * @param allKeys   — all config keys (flat)
 * @param patterns  — glob-like patterns (e.g., ["*_KEY", "*_SECRET", "DATABASE_*"])
 * @param schemaSensitive — keys explicitly marked sensitive in schema
 * @returns Set of matching sensitive keys
 */
export function buildSensitiveSetWithPatterns(
  allKeys: string[],
  patterns: string[],
  schemaSensitive: Set<string> = new Set(),
): Set<string> {
  const result = new Set(schemaSensitive);
  const compiled = compilePatterns(patterns);

  for (const key of allKeys) {
    const leaf = key.includes(".") ? key.split(".").pop()! : key;
    // Check against custom patterns
    if (compiled.some((re) => re.test(leaf) || re.test(key))) {
      result.add(key);
    }
    // Also check built-in patterns
    if (isAutoSensitive(leaf)) {
      result.add(key);
    }
  }

  return result;
}

/** Callback for secret access audit logging. */
export type SecretAccessCallback = (key: string, context: string) => void;

/**
 * Create a Proxy that wraps a config object and fires audit callbacks
 * when sensitive keys are accessed.
 *
 * @param data            — the config data
 * @param sensitiveKeys   — set of sensitive keys
 * @param onAccess        — callback fired each time a sensitive key is read
 * @returns Proxied config object
 */
export function createAuditProxy(
  data: Record<string, unknown>,
  sensitiveKeys: Set<string>,
  onAccess: SecretAccessCallback,
): Record<string, unknown> {
  return new Proxy(data, {
    get(target, prop, receiver) {
      if (typeof prop === "string" && sensitiveKeys.has(prop)) {
        const context = typeof (globalThis as Record<string, unknown>).window !== "undefined" ? "client" : "server";
        onAccess(prop, context);
      }
      return Reflect.get(target, prop, receiver);
    },
  });
}

